src/cascade/executor/create_settings.py

```diff
@@ -211,45 +258,100 @@ def create_settings(rng, locations):
-        if scovariates:
-            case[f"{ckind}_covariate"] = scovariates
+def add_chosen_random_effects(case, location_root, locations, rate_specifies_re_by_location, rng):
+    random_effects = list()
+    # Iterates over every set of parent and children in the graph.
+    for rate_name, by_location in rate_specifies_re_by_location:
+        if by_location:
+            # You have to do _all_ of them.
+            for parent, children in nx.bfs_successors(locations, location_root):
+                random_effects.extend(add_random_effects(children, rate_name, rng))
+        else:
+            add_random_effects(None, rate_name, rng)
+    # Only add to dict if there are some?
+    if random_effects:
+        case["random_effect"] = random_effects
 
 
-def add_random_effects(children, has, rate, rng):
-    random_effects = list()
-    if children:
-        for random_effect, exists in [(x, y) for (x, y) in has.items() if y]:
-            single_age = rate == "pini"
-            choice = rng.choice(["none", "all", "every"])
-            if choice == "all":
-                grid_case = grid(False, rng, single_age)
-                grid_case["rate"] = random_effect
-                grid_case["location"] = 1
-                random_effects.append(grid_case)
-            elif choice == "every":
-                for child in children:
-                    grid_case = grid(False, rng, single_age)
-                    grid_case["rate"] = random_effect
-                    grid_case["location"] = child
-                    random_effects.append(grid_case)
-    return random_effects
+def which_random_effects(nonzero_rates, rate, rng):
+    # We specify priors of random effects either
+    # a) Once for the whole hierarchy, or
+    # b) For every single location in the hierarchy.
+    # Choose this by rate
+    rate_specifies_re_by_location = dict()
+    for rate_name in nonzero_rates:
+        choice = rng.choice(["none", "all", "every"], name=f"re.{rate_name}")
+        # none means no random effects.
+        # all means every single location.
+        # every means each location gets a new one.
+        if rate in {"all", "every"}:
+            rate_specifies_re_by_location[rate_name] = choice == "every"
+        else:
+            pass  # Don't record this rate as needing a random effect.
+    return rate_specifies_re_by_location
+
+
+class SettingsChoices:
+    """
+    This class will fix certain choices to given values. Its input is
+    a list of key-value pairs, set up as for a section of a ConfigParser,
+    although this adds the section title for you.
+    """
+    def __init__(self, rng=None, settings=None):
+        if isinstance(settings, str):
+            parser = ConfigParser()
+            parser.read_string("[settings]\n" + dedent(settings))
+            self.answers = parser["settings"]
+        else:
+            self.answers = dict()
+        self.rng = rng
```

Comment:
What would it mean for the rng to be None? It seems to me that if the user doesn't pass in their own randomstate we should make one for them.